File: frontend/src/types/collaboration.ts
export type CollaborationAction = "viewing" | "editing" | "idle";

export interface EditingArea {
  row_key?: string;
  field?: string;
  node_id?: string;
  section?: string;
  column?: string;
}

export interface ActiveUser {
  user_id: string;
  user_name: string;
  action: CollaborationAction;
  editing_area: EditingArea | null;
}

export interface CollaborationState {
  activeUsers: ActiveUser[];
  currentUserEditing: boolean;
  isSyncing: boolean;
  startEditing: (area: EditingArea) => void;
  stopEditing: () => void;
}

export interface ConflictInfo {
  saved_by: string | null;
  saved_at: string | null;
  latest_lock_version: number;
}

File: backend/app/schemas/collaboration.py
import uuid
from typing import Literal

from pydantic import BaseModel


class EditingArea(BaseModel):
    row_key: str | None = None
    field: str | None = None
    node_id: str | None = None
    section: str | None = None
    column: str | None = None


class HeartbeatRequest(BaseModel):
    document_type: str
    document_id: uuid.UUID
    action: Literal["viewing", "editing", "idle"] = "viewing"
    editing_area: EditingArea | None = None


class ActiveUser(BaseModel):
    user_id: str
    user_name: str
    action: Literal["viewing", "editing", "idle"]
    editing_area: EditingArea | None = None


class ActiveUsersResponse(BaseModel):
    users: list[ActiveUser]
    total: int
